apps/frontend/src/pages/CreateCharity.tsx

```diff
@@ -0,0 +1,651 @@
+    const uploadPromises = Images.map(async (image) => {
+      try {
+        const imageFile = image.image;
+        // Upload the image
+        const response = await _axios_instance.post("/images", imageFile, {
+          headers: {
+            "Content-Type": "multipart/form-data",
+          },
+        });
+
+        return { url: response.data.url, index: image.index };
+      } catch (error) {
+        console.error("Error uploading image:", error);
+        return null;
+      }
+    });
+
+    try {
+      const results = await Promise.all(uploadPromises);
+
+      results.forEach((result) => {
+        if (result) {
+          retrievedImages.push(result);
+        }
+      });
+
+      // Return the list of successfully uploaded images
+      return retrievedImages;
+    } catch (error) {
+      console.error("Error in uploading image process:", error);
+      return false;
+    }
+  };
+
+  // Uploads the main charity image to the s3 blob store
+  const asyncListingImageWrapperLogo = async (): Promise<boolean | string> => {
+    try {
+      if (!logoImage) return true;
+      const logoImagesObjectArray = await asyncUploadImages([
+        { image: logoImage, index: 0 },
+      ]);
+
+      if (logoImagesObjectArray) {
+        const newURL = logoImagesObjectArray[0].url;
+        setNewCharityObject((prevState) => ({
+          ...prevState,
+          imageUrl: newURL,
+        }));
+
+        return newURL;
+      } else {
+        console.error("No images uploaded for logoImage.");
+        return false;
+      }
+    } catch (error) {
+      console.error("Error uploading logo image:", error);
+      return false;
+    }
+  };
+
+  // Uploads the partner images to the s3 blobstore and packages the partner logos
+  const asyncListingImageWrapperPartners = async (
+    logoUrlResult: string | boolean,
+  ): Promise<boolean | CharityObject> => {
+    try {
+      if (partnerImages.length === 0) {
+        return { ...newCharityObject, imageUrl: logoUrlResult as string };
+      }
+
+      // Check if newCharityObject.organizations is an array
+      if (!Array.isArray(newCharityObject.organizations)) {
+        console.error(
+          "newCharityObject.organizations is not an array:",
+          newCharityObject.organizations,
+        );
+        return false;
+      }
+
+      // Upload partnerImages
+      const partnerImagesArray = partnerImages
+        .filter((obj) => obj.image !== undefined)
+        .map((obj, index) => ({ image: obj.image as File, index }));
+
+      const partnerImagesObjectArray =
+        await asyncUploadImages(partnerImagesArray);
+      if (partnerImagesObjectArray && partnerImagesObjectArray.length > 0) {
+        const newOrganizations = [...newCharityObject.organizations];
+        for (let i = 0; i < newCharityObject.organizations.length; i++) {
+          newOrganizations[partnerImagesObjectArray[i].index].logoUrl =
+            partnerImagesObjectArray[i].url;
+        }
```

Comment:
In practice JavaScript doesn't change the order of an array on its own (as far as I know / have encountered), but the docs specifically state that **array order is not guaranteed**. An implementation that creates organizations one by one rather than dealing with an array of images and indexing them all might be safer and also more simple, avoiding these for loops with fancy indexing (presumably `partherImagesObjectArray[i].index` should just be i?)

Reply:
Yeah, I fiddled with a couple ways to do this, but basically the organizations all need to be attached to a specific image, and the requests aren't guarenteed to be resolved in order, so I need to attach them to the index of the organization that called them. I tried to attach them directly to the organizations when submitting but the object never updated when I needed it to and this was working